tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc

```diff
@@ -4190,25 +4188,108 @@ tsl::StatusOr<dnn::DataType> PrimitiveTypeStringToDnnType(
+  }
+
+  KINDS_AND_MODE_FROM_OP_STRING("add", TensorKind::kTensor, TensorKind::kTensor,
+                                CUDNN_POINTWISE_ADD)
+  KINDS_AND_MODE_FROM_OP_STRING("relu", TensorKind::kNone, TensorKind::kTensor,
+                                CUDNN_POINTWISE_RELU_FWD)
+  KINDS_AND_MODE_FROM_OP_STRING("scale", TensorKind::kScalar,
+                                TensorKind::kTensor, CUDNN_POINTWISE_MUL)
+  KINDS_AND_MODE_FROM_OP_STRING("invscale", TensorKind::kScalar,
+                                TensorKind::kTensor, CUDNN_POINTWISE_DIV)
+  KINDS_AND_MODE_FROM_OP_STRING("amax", TensorKind::kNone, TensorKind::kScalar,
+                                CUDNN_REDUCE_TENSOR_AMAX)
+#undef KINDS_AND_MODE_FROM_OP_STRING
 
   return tsl::errors::Internal("Unknown op.");
 }
 
+// Struct describing the convolution, pointwise and reduction ops in the
+// graph.
+struct OpDescriptor {
+  int uid;                         // The UID of the op.
+  std::optional<int> operand_uid;  // The UID of the at most one operand of the
+                                   // op which is part of the graph.
+  OpMode mode;                     // The mode describing the op.
+  TensorKind operand_kind;    // The kind of a second operand (side input) not
+                              // represented in the graph.
+  TensorKind result_kind;     // The kind of the output.
+  dnn::DataType result_type;  // The type of the output.
+  bool is_virtual;            // A virtual op has a user within the graph.
+  int sequence_index;         // The index of the op in the sequence.
+};
+
+// Class describing the graph of ops to be fused into the cuDNN convolution
+// Custom Call.
+class OpGraph {
+ public:
+  OpGraph() : ops_index_(0){};
+
+  tsl::Status AddOp(int uid, std::optional<int> operand_uid, OpMode mode,
+                    TensorKind operand_kind, TensorKind result_kind,
+                    dnn::DataType result_type) {
+    ops_.emplace_back(OpDescriptor({uid, operand_uid, mode, operand_kind,
+                                    result_kind, result_type, false, -1}));
+    // If it exists, the operand is virtual.
+    if (operand_uid.has_value()) {
+      auto it = std::find_if(
+          ops_.begin(), ops_.end(),
+          [operand_uid](OpDescriptor op) { return op.uid == operand_uid; });
+      if (it == ops_.end()) {
+        return tsl::errors::Internal("Unknown ID.");
+      }
+      it->is_virtual = true;
+    }
+    return tsl::OkStatus();
+  }
+
+  tsl::StatusOr<OpDescriptor> FindOpDescriptor(int uid) const {
+    auto it = std::find_if(ops_.begin(), ops_.end(),
+                           [uid](OpDescriptor op) { return op.uid == uid; });
+    if (it == ops_.end()) {
+      return tsl::errors::Internal("Unknown ID.");
+    }
+    return *it;
+  }
+
+  std::optional<OpDescriptor> NextOpDescriptor() {
+    if (ops_.size() > ops_index_) {
+      return ops_[ops_index_++];
+    }
+    return std::nullopt;
+  }
```

Comment:
I don't think we should store `ops_index_` in this class and have this method, since it's combining the OpGraph class with an iterator over the graph. Instead have a `GetOpAtIndex(int index)` function.